haas/ext/auth/keystone.py

```diff
@@ -0,0 +1,64 @@
+"""Keystone authentication backend.
+
+This is a thin wrapper around the `keystonemiddleware` library.
+"""
+from keystonemiddleware.auth_token import filter_factory
+from flask import request
+from haas.flaskapp import app
+from haas.config import cfg
+from haas.model import Project
+from haas import auth, rest
+import logging
+import sys
+
+logger = rest.ContextLogger(logging.getLogger(__name__), {})
+
+
+class KeystoneAuthBackend(auth.AuthBackend):
+
+    def authenticate(self):
+        # keystonemiddleware makes auth info available from two places:
+        #
+        # 1. Variables in the wsgi environment
+        # 2. Extra HTTP headers.
+        #
+        # In general, the wsgi variable 'HTTP_FOO_BAR' is equivalent to the
+        # HTTP header 'Foo-Bar'.
+        #
+        # We use the wsgi environment's variables below; it shouldn't matter,
+        # but this way if something goes horribly wrong and arbitrary headers
+        # aren't stripped out, the client can't just inject these.
+        if request.environ['HTTP_X_IDENTITY_STATUS'] != 'Confirmed':
+            return False
+
+        project_id = request.environ['HTTP_X_PROJECT_ID']
+        if Project.query.filter_by(label=project_id).count() == 0:
+            logger.info("Successful authentication by Openstack project %r, "
+                        "but this project is not registered with HaaS",
+                        project_id)
+            return False
+
+        return True
+
+    def _have_project_access(self, project):
+        return project.label == request.environ['HTTP_X_PROJECT_ID']
+
+    def _have_admin(self):
+        return 'admin' in request.environ['HTTP_X_ROLES'].split(',')
```

Comment:
@henn It depends on the Keystone configuration and policy file in use. With RBAC it is now possible to write policy files which restrict and scope the global admin. See [bug 968696](https://bugs.launchpad.net/keystone/+bug/968696)

`HTTP_X_IS_ADMIN_PROJECT` requires setting the configuration values for `admin_project_name` and `admin_project_domain_name`, which default to `None`. [[Source]](https://github.com/openstack/keystone/blob/master/etc/keystone.conf.sample#L1760)

Also the new sample policy configuration for v3 makes use of this new `cloud_admin` role [[Source]](https://github.com/openstack/keystone/blob/master/etc/policy.v3cloudsample.json#L3), but the default one does not [[Source]](https://github.com/openstack/keystone/blob/master/etc/policy.json#L2).